src/main/java/org/jabref/gui/strings/StringDialogViewModel.java

```diff
@@ -0,0 +1,85 @@
+package org.jabref.gui.strings;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javafx.beans.property.ListProperty;
+import javafx.beans.property.ObjectProperty;
+import javafx.beans.property.SimpleListProperty;
+import javafx.beans.property.SimpleObjectProperty;
+import javafx.beans.property.SimpleStringProperty;
+import javafx.beans.property.StringProperty;
+import javafx.collections.FXCollections;
+
+import org.jabref.gui.AbstractViewModel;
+import org.jabref.logic.bibtex.comparator.BibtexStringComparator;
+import org.jabref.model.database.BibDatabase;
+import org.jabref.model.entry.BibtexString;
+
+public class StringDialogViewModel extends AbstractViewModel {
+
+    private final ListProperty<StringViewModel> allStrings = new SimpleListProperty<>(FXCollections.observableArrayList());
+    private final ObjectProperty<StringViewModel> selectedItemProperty = new SimpleObjectProperty<>();
+    private final StringProperty newStringLabelProperty = new SimpleStringProperty("");
+    private final BibDatabase db;
+
+    public StringDialogViewModel(BibDatabase bibDatabase) {
+        this.db = bibDatabase;
+        addAllStringsFromDB();
+
+    }
+
+    private void addAllStringsFromDB() {
+
+        Set<StringViewModel> strings = db.getStringKeySet()
+                                         .stream()
+                                         .map(string -> db.getString(string))
+                                         .sorted(new BibtexStringComparator(false))
+                                         .map(this::convertFromBibTexString)
+                                         .collect(Collectors.toSet());
+        allStrings.addAll(strings);
+
+    }
+
+    public ListProperty<StringViewModel> allStringsProperty() {
+        return this.allStrings;
+    }
+
+    public void addNewString() {
+        allStrings.add(new StringViewModel(newStringLabelProperty.getValue(), "New Content"));
+    }
+
+    public void removeString() {
+        allStrings.remove(selectedItemProperty.getValue());
+    }
+
+    private StringViewModel convertFromBibTexString(BibtexString bibtexString) {
+        return new StringViewModel(bibtexString.getName(), bibtexString.getContent());
+    }
+
+    public ObjectProperty<StringViewModel> seletedItemProperty() {
+        return this.selectedItemProperty;
+    }
+
+    public StringProperty newStringLabelProperty() {
+        return this.newStringLabelProperty;
+    }
+
+    public void save() {
+        for (StringViewModel model : allStrings) {
```

Comment:
move this to a new method `BibDatabase.setStrings(Collection<BibtexString>))` ?

Reply:
Not that easy, it's a HashMap which explicitly checks the key before inserting and generates a new id.

  ```
  /**
     * Inserts a Bibtex String.
     */
    public synchronized void addString(BibtexString string) throws KeyCollisionException {
        if (hasStringLabel(string.getName())) {
            throw new KeyCollisionException("A string with that label already exists");
        }

        if (bibtexStrings.containsKey(string.getId())) {
            throw new KeyCollisionException("Duplicate BibTeX string id.");
        }

        bibtexStrings.put(string.getId(), string);
    }


```

Reply:
I meant you should simply refactor the code and move (most) of it to a new method in `BibDatabase` (not re-implement it in another way).